File: Excel/900-999/980/Challenge 980.py
import pandas as pd
from itertools import permutations

path = "900-999/980/980 Max and Min Numbers.xlsx"
input = pd.read_excel(path, usecols="A:E", nrows=11, skiprows=1, dtype=str)
test = pd.read_excel(path, usecols="F:G", nrows=11, skiprows=1, dtype=str)

long_df = (
	input.assign(rn=range(1, len(input) + 1))
	.melt(id_vars="rn", value_name="value")
	.dropna(subset=["value"])
)
long_df["value"] = long_df["value"].astype(str).str.replace(r"0\.", ".", regex=True)
rows = []
for rn, grp in long_df.groupby("rn", sort=False):
	values = grp["value"].tolist()
	combinations = {"".join(p) for p in permutations(values, len(values))}
	comb_df = pd.DataFrame({"rn": rn, "combination": list(combinations)})
	comb_df["num_combination"] = pd.to_numeric(comb_df["combination"], errors="coerce")
	comb_df = comb_df.dropna(subset=["num_combination"])
	if comb_df.empty:
		continue
	min_val, max_val = comb_df["num_combination"].agg(["min", "max"])
	mm = comb_df[comb_df["num_combination"].isin([min_val, max_val])].copy()
	mm["minmax"] = mm["num_combination"].map(
		lambda x: "MIN" if x == min_val else "MAX"
	)
	rows.append(mm[["rn", "combination", "minmax"]])
if rows:
	result = (
		pd.concat(rows, ignore_index=True)
		.drop_duplicates()
		.pivot_table(index="rn", columns="minmax", values="combination", aggfunc="first")
		.reset_index(drop=True)
		.reindex(columns=["MIN", "MAX"])
	)
	result.columns.name = None
else:
	result = pd.DataFrame(columns=["MIN", "MAX"])

# Some discrepancies.

File: Excel/900-999/980/Challenge 980.R
library(tidyverse)
library(readxl)
library(gtools)

path <- "900-999/980/980 Max and Min Numbers.xlsx"
input <- read_excel(path, range = "A2:E12", col_types = "text")
test <- read_excel(path, range = "F2:G12")

options(scipen = 999)

result = input %>%
  mutate(rn = row_number()) %>%
  pivot_longer(-rn, names_to = "type", values_to = "value") %>%
  na.omit() %>%
  mutate(value = str_replace(value, "0\\.", "\\.")) %>%
  summarise(
    combination = list(apply(
      permutations(length(value), length(value), seq_along(value)),
      1,
      \(i) str_c(value[i], collapse = "")
    )),
    .by = rn
  ) %>%
  unnest(combination) %>%
  mutate(
    n = as.numeric(combination),
    minmax = case_when(n == min(n) ~ "MIN", n == max(n) ~ "MAX"),
    .by = rn
  ) %>%
  filter(!is.na(minmax)) %>%
  distinct() %>%
  select(-n) %>%
  pivot_wider(names_from = minmax, values_from = combination) %>%
  select(-rn)

result == test
as.numeric(result$MAX) > as.numeric(test$MAX)
as.numeric(result$MIN) < as.numeric(test$MIN)
# some discrepancies.
